typing/typedecl.ml

```diff
@@ -1263,8 +1263,28 @@ let name_recursion sdecl id decl =
     else decl
   | _ -> decl
 
+(* Warn on definitions of type "type foo = ()" which redefine a different unit
+   type and are likely a mistake. *)
+let check_redefined_unit (td: Parsetree.type_declaration) =
+  let open Parsetree in
+  let is_unit_constructor : constructor_declaration -> bool = function
+    | { pcd_name = { txt = "()" };
+        pcd_args = Pcstr_tuple [];
```

Comment:
I am not sure if it is useful to not warn on `type t = ( ) of int `.

Reply:
My initial reasoning was that if you're doing something very strange like this, then there must be a reason.. In the other hand, if we can't think of a reason, then we might as well warn on it indeed.

Reply:
I amended the code to also warn on these cases (and also stuff like `type _ t = () : int foo` :-).